oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/NodeDocumentCodec.java

```diff
@@ -0,0 +1,161 @@
+    // The estimated size is stored in the NodeDocument itself
+    public final static String SIZE_FIELD = "__ESTIMATED_SIZE__";
+    private final MongoDocumentStore store;
+    private final Collection<NodeDocument> collection;
+    private final BsonTypeCodecMap bsonTypeCodecMap;
+    private final DecoderContext decoderContext = DecoderContext.builder().build();
+
+    private final Codec<String> stringCoded;
+    private final Codec<Long> longCoded;
+    private final Codec<Boolean> booleanCoded;
+
+    private int estimatedSizeOfCurrentObject = 0;
+
+    public NodeDocumentCodec(MongoDocumentStore store, Collection<NodeDocument> collection, CodecRegistry defaultRegistry) {
+        this.store = store;
+        this.collection = collection;
+        this.bsonTypeCodecMap = new BsonTypeCodecMap(new BsonTypeClassMap(), defaultRegistry);
+        // Retrieve references to the most commonly used codecs, to avoid the map lookup in the common case
+        this.stringCoded = (Codec<String>) bsonTypeCodecMap.get(BsonType.STRING);
+        this.longCoded = (Codec<Long>) bsonTypeCodecMap.get(BsonType.INT64);
+        this.booleanCoded = (Codec<Boolean>) bsonTypeCodecMap.get(BsonType.BOOLEAN);
+    }
+
+    @Override
+    public NodeDocument decode(BsonReader reader, DecoderContext decoderContext) {
+        NodeDocument nodeDocument = collection.newDocument(store);
+        estimatedSizeOfCurrentObject = 0;
+        reader.readStartDocument();
+        while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
+            String fieldName = reader.readName();
+            Object value = readValue(reader, fieldName);
+            nodeDocument.put(fieldName, value);
+        }
+        reader.readEndDocument();
+        nodeDocument.put(SIZE_FIELD, estimatedSizeOfCurrentObject);
+        return nodeDocument;
+    }
+
+    @Override
+    public void encode(BsonWriter writer, NodeDocument value, EncoderContext encoderContext) {
+        throw new UnsupportedOperationException("encode");
+    }
+
+    @Override
+    public Class<NodeDocument> getEncoderClass() {
+        return NodeDocument.class;
+    }
+
+    private Object readValue(BsonReader reader, String fieldName) {
+        estimatedSizeOfCurrentObject += 16 + fieldName.length();
+        BsonType bsonType = reader.getCurrentBsonType();
+        Object value;
+        int valSize;
+        switch (bsonType) {
+            case STRING:
+                String sValue = stringCoded.decode(reader, decoderContext);
+                valSize = 16 + sValue.length() * 2;
+                value = sValue;
+                break;
+            case INT64:
+                value = longCoded.decode(reader, decoderContext);
+                valSize = 16;
+                break;
+            case DOCUMENT:
+                value = readDocument(reader);
+                valSize = 0; // the size is updated by the recursive calls inside convertMongoMap
```

Comment:
I think you mean `readDocument` here
```suggestion
                valSize = 0; // the size is updated by the recursive calls inside readDocument
```